File: 3720-lexicographically-smallest-permutation-greater-than-target/solution.py
class Solution:
    def nextPermutation(self, s: str, target: str) -> str:
        n = len(s)
        counts = [0] * 26
        
        # Count frequencies of characters in s
        for char in s:
            counts[ord(char) - ord('a')] += 1
            
        # Find maximum prefix length L of target that s can match
        L = 0
        while L < n and counts[ord(target[L]) - ord('a')] > 0:
            counts[ord(target[L]) - ord('a')] -= 1
            L += 1
            
        # If target was completely matched, adjust L to n - 1 and restore target[n - 1]
        if L == n:
            L -= 1
            counts[ord(target[L]) - ord('a')] += 1
            
        # Try to find the largest index i where we can place a character > target[i]
        for i in range(L, -1, -1):
            target_char_idx = ord(target[i]) - ord('a')
            
            # Look for the smallest available character > target[i]
            replacement_char = -1
            for c in range(target_char_idx + 1, 26):
                if counts[c] > 0:
                    replacement_char = c
                    break
                    
            if replacement_char != -1:
                # Found a valid configuration! Construct the answer string
                res = list(target[:i])
                
                # Place the strictly greater character
                res.append(chr(ord('a') + replacement_char))
                counts[replacement_char] -= 1
                
                # Fill the rest with remaining characters in ascending order
                for c in range(26):
                    while counts[c] > 0:
                        res.append(chr(ord('a') + c))
                        counts[c] -= 1
                        
                return "".join(res)
                
            # Backtrack target[i - 1] for index i - 1
            if i > 0:
                counts[ord(target[i - 1]) - ord('a')] += 1
                
        return ""

File: 3720-lexicographically-smallest-permutation-greater-than-target/solution.cpp
#include <string>
#include <vector>
#include <numeric>

class Solution {
public:
    string nextPermutation(string s, string target) {
        int n = s.length();
        vector<int> counts(26, 0);
        
        // Count frequencies of characters in s
        for (char c : s) {
            counts[c - 'a']++;
        }
        
        // Find maximum prefix length L of target that s can match
        int L = 0;
        while (L < n && counts[target[L] - 'a'] > 0) {
            counts[target[L] - 'a']--;
            L++;
        }
        
        // If target was completely matched, adjust L to n - 1 and restore target[n - 1]
        if (L == n) {
            L--;
            counts[target[L] - 'a']++;
        }
        
        // Try to find the largest index i where we can place a character > target[i]
        for (int i = L; i >= 0; --i) {
            int targetCharIndex = target[i] - 'a';
            
            // Look for the smallest available character > target[i]
            int replacementChar = -1;
            for (int c = targetCharIndex + 1; c < 26; ++c) {
                if (counts[c] > 0) {
                    replacementChar = c;
                    break;
                }
            }
            
            if (replacementChar != -1) {
                // Found a valid configuration! Construct the answer string
                string result = target.substr(0, i);
                
                // Place the strictly greater character
                result += (char)('a' + replacementChar);
                counts[replacementChar]--;
                
                // Fill the rest with remaining characters in ascending order
                for (int c = 0; c < 26; ++c) {
                    while (counts[c] > 0) {
                        result += (char)('a' + c);
                        counts[c]--;
                    }
                }
                
                return result;
            }
            
            // If index i cannot be made larger, backtrack target[i - 1] for index i - 1
            if (i > 0) {
                counts[target[i - 1] - 'a']++;
            }
        }
        
        return "";
    }
};